src/commands/dockerInstaller.ts

```diff
@@ -12,27 +13,47 @@ import { localize } from '../localize';
 import { dockerInstallStatusProvider } from '../utils/DockerInstallStatusProvider';
 import { streamToFile } from '../utils/httpRequest';
 import LocalOSProvider from '../utils/LocalOSProvider';
+import { openExternal } from '../utils/openExternal';
 import { execAsync } from '../utils/spawnAsync';
 
 export abstract class DockerInstallerBase {
     protected abstract downloadUrl: string;
     protected abstract fileExtension: string;
     protected abstract getInstallCommand(fileName: string): string;
 
-    public async downloadAndInstallDocker(): Promise<void> {
+    public async downloadAndInstallDocker(context: IActionContext): Promise<void> {
         const shouldInstall: boolean = await this.preInstallCheck();
         if (shouldInstall) {
             const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
             const installationMessage: string = localize('vscode-docker.commands.DockerInstallerBase.installationMessage', 'The Docker Desktop installation is started. Complete the installation and then start Docker Desktop.');
+            let downloadedFileName: string;
 
-            const downloadedFileName: string = await vscode.window.withProgress(
-                { location: vscode.ProgressLocation.Notification, title: downloadingMessage },
-                async () => this.downloadInstaller());
+            context.telemetry.properties.stage = 'download';
+            try {
+                downloadedFileName = await vscode.window.withProgress(
+                    { location: vscode.ProgressLocation.Notification, title: downloadingMessage },
+                    async () => this.downloadInstaller());
+            } catch (error) {
+                const message = localize('vscode-docker.commands.DockerInstallerBase.downloadFailed', 'Downloading the Docker Desktop installer failed. Do you want to manually download and install?');
+                const title = localize('vscode-docker.commands.DockerInstallerBase.download', 'Download');
+                const response = await vscode.window.showErrorMessage(message, { title: title });
+                if (response) {
+                    await openExternal(this.downloadUrl);
+                }
+                context.errorHandling.suppressReportIssue = true;
+                context.errorHandling.suppressDisplay = true;
+                throw error;
+            }
 
+            context.telemetry.properties.stage = 'install';
             const command = this.getInstallCommand(downloadedFileName);
             // eslint-disable-next-line @typescript-eslint/no-floating-promises
             vscode.window.showInformationMessage(installationMessage);
-            await this.install(downloadedFileName, command);
+            try {
+                await this.install(downloadedFileName, command);
```

Comment:
We should wrap with an error message that provides context. Otherwise they might get just "Process exited with code X" on Windows, which isn't very helpful out-of-context.

Reply:
We can also do this to avoid duplicating `showErrorMessage`:

```typescript
let message: string;
let button: string;

try {
    message = localize(...);
    title = localize(...);
    // ... download

    message = localize(...);
    button = localize(...);
    // ... install
} catch {
    vscode.window.showErrorMessage(...).then(...);
}
```

Reply:
Added something similar